### PR TITLE
ml - Save Image Link

### LINKED ISSUE #2: Save Image Link
#### As a user I want to save a link to an image.

_Assignees_: @HaominZhang, @XanderStrike

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [x] The landing page links to the image link submission form.
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.
- [x] After the form is submitted, the user is redirected to a page displaying
  the image.
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.

__Discussion Topic__:
- [x] What is a valid image URL?

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

### PR DESCRIPTION
**Pull request by mluders**
Closes #2 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [ ] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: app/models/image.rb [hunk 1]
```diff
@@ -1,7 +1,19 @@
...
                     length: { minimum: 5 }
+
+  private
+
+  def acceptable_file_extension?
+    valid = false
+    %w[.jpeg .png .gif].each do |ext|
+      valid = true if url.ends_with?(ext)
+    end
+    errors.add(:url, ' is not an acceptable type. Must be .jpeg, .png, or .gif.') unless valid
```
Comment (HaominZhang) [inline app/models/image.rb]: 👍

File: app/views/images/index.html.erb [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
-    <%= link_to image_path(image) do %>
-      <li>
-        <image src="<%= image.url %>"></image>
-        <p><%= image.title %></p>
-      </li>
-    <% end %>
-  </ul>
+<%= link_to 'Add an image', new_image_path %>
 
+<% @images.each do |image| %>
```
Comment (HaominZhang) [inline app/views/images/index.html.erb]: Not a big deal, but I think this part belongs to next issue. How about we commit this change later?

File: db/migrate/20180718001708_create_images.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class CreateImages < ActiveRecord::Migration[5.2]
+  def change
+    create_table :images do |t|
+      t.text :url
```
Comment (HaominZhang) [inline db/migrate/20180718001708_create_images.rb]: Do we want this url to be blank on database level? If not, we could add `null: false` restriction

File: app/controllers/images_controller.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+  def new
+    @image = Image.new
+  end
+
+  def create
+    @image = Image.new(image_params)
+
+    if @image.save
+      redirect_to @image
```
Comment (HaominZhang) [inline app/controllers/images_controller.rb]: Do we want to render any flash messages to tell user if they action succeeded or not?

File: db/migrate/20180718001708_create_images.rb (hunk 3 above)
Comment (mluders) [reply db/migrate/20180718001708_create_images.rb]: Good idea

File: app/controllers/images_controller.rb (hunk 4 above)
Comment (mluders) [reply app/controllers/images_controller.rb]: Good idea as well. I'll implement that.

**Pull request merged**